assets/js/base/context/providers/cart-checkout/payment-methods/payment-method-data-context.tsx

```diff
@@ -90,103 +62,58 @@ export const PaymentMethodDataProvider = ( {
 		isCalculating: checkoutIsCalculating,
 		hasError: checkoutHasError,
 	} = useCheckoutContext();
 	const { isEditor, getPreviewData } = useEditorContext();
+	const { setValidationErrors } = useValidationContext();
+	const { addErrorNotice, removeNotice } = useStoreNotices();
 	const {
 		isSuccessResponse,
 		isErrorResponse,
 		isFailResponse,
 		noticeContexts,
 	} = useEmitResponse();
-	const { dispatchCheckoutEvent } = useStoreEvents();
-
-	const [ activePaymentMethod, setActive ] = useState( '' ); // The active payment method - e.g. Stripe CC or BACS.
-	const [ activeSavedToken, setActiveSavedToken ] = useState( '' ); // If a previously saved payment method is active, the token for that method. For example, a for a Stripe CC card saved to user account.
 	const [ observers, observerDispatch ] = useReducer( emitReducer, {} );
+	const { onPaymentProcessing } = useEventEmitters( observerDispatch );
+	const currentObservers = usePrevious( observers );
+
 	const [ paymentData, dispatch ] = useReducer(
 		reducer,
 		DEFAULT_PAYMENT_DATA_CONTEXT_STATE
 	);
-	const currentObservers = useRef( observers );
-	const { onPaymentProcessing } = useEventEmitters( observerDispatch );
-
-	// ensure observers are always current.
-	useEffect( () => {
-		currentObservers.current = observers;
-	}, [ observers ] );
-
-	const setActivePaymentMethod = useCallback(
-		( paymentMethodSlug ) => {
-			setActive( paymentMethodSlug );
-			dispatch( statusOnly( PRISTINE ) );
-			dispatchCheckoutEvent( 'set-active-payment-method', {
-				paymentMethodSlug,
-			} );
-		},
-		[ setActive, dispatch, dispatchCheckoutEvent ]
-	);
-
-	const paymentMethodsDispatcher = useCallback<
-		PaymentMethodsDispatcherType
-	>(
-		( paymentMethods ) => {
-			dispatch(
-				setRegisteredPaymentMethods( paymentMethods as PaymentMethods )
-			);
-		},
-		[ dispatch ]
-	);
-
-	const expressPaymentMethodsDispatcher = useCallback<
-		PaymentMethodsDispatcherType
-	>(
-		( paymentMethods ) => {
-			dispatch(
-				setRegisteredExpressPaymentMethods(
-					paymentMethods as PaymentMethods
-				)
-			);
-		},
-		[ dispatch ]
-	);
+	const {
+		dispatchActions,
+		setPaymentStatus,
+	} = usePaymentMethodDataDispatchers( dispatch );
 
 	const paymentMethodsInitialized = usePaymentMethods(
-		paymentMethodsDispatcher
+		dispatchActions.setRegisteredPaymentMethods
 	);
 
 	const expressPaymentMethodsInitialized = useExpressPaymentMethods(
-		expressPaymentMethodsDispatcher
+		dispatchActions.setRegisteredExpressPaymentMethods
 	);
 
-	const { setValidationErrors } = useValidationContext();
-	const { addErrorNotice, removeNotice } = useStoreNotices();
-	const { setShippingAddress } = useShippingDataContext();
-	const setShouldSavePayment = useCallback(
-		( shouldSave ) => {
-			dispatch( setShouldSavePaymentMethod( shouldSave ) );
-		},
-		[ dispatch ]
-	);
+	const {
+		activePaymentMethod,
+		activeSavedToken,
+		setActivePaymentMethod,
+		setActiveSavedToken,
+	} = useActivePaymentMethod();
```

Comment:
Also split the active method tracking into custom hook to reduce the size of this function. Logic is the same.